analytics/core.go

```diff
@@ -35,6 +35,8 @@ type AuctionObject struct {
 	Account              *config.Account
 	StartTime            time.Time
 	HookExecutionOutcome []hookexecution.StageOutcome
+	SeatNonBid           []openrtb_ext.SeatNonBid    `json:"-"`
+	RequestWrapper       *openrtb_ext.RequestWrapper `json:"-"`
```

Comment:
I empathize with wanting to avoid unmarshalling extension objects within each analytics adapter. We need to balance with the danger of memory corruption since the RequestWrapper is not thread safe and even the Get methods modify memory (so this is a different situation than enforcing no writes to the Request object). There is now the ability to clone a RequestWrapper , so we could clone it before passing it to each adapter - but then we need to consider if the cost of memory copying is worth the unmarshal savings.

If we replace the Request with the RequestWrapper the json marshal would change. However, I am thinking now that the file system and pubstack adapters make a mistake by using the data objects in this way. We could solve that problem by creating separate objects for each of those adapters to isolate the changes. This would also remove the need to specify  `json:"-"`.

Thoughts @hhhjort @bsardo ?

Reply:
I opened https://github.com/prebid/prebid-server/pull/2739 to separate the data models so you don't need to add `json:"-"` here and if we decide to replace Request with RequestWrapper, the analytics adapters can handle the change.

Reply:
If analytics adapters are only going to use the wrapper to read the extension, and not write anything, then it should be fairly safe to pass them the wrapper. The only thing that could happen on read is on the GetXExt() call where you get a wrapper Ext object. If one doesn't already exist on the wrapper, the wrapper will spin up a new one to return. If two adapters call this at the same time, you could get two copies of the ext object. The copies will be identical clones, and each adapter will consistently use the one it started with. After the race condition window, the wrapper will only have one of them. The other copy will persist until the adapter finishes with it and it is garbage collected.

So as long as we police the analytics adapters to not write to the request (which we should do anyway), and PBS core is not modifying the request wrapper while the analytics adapters are doing their thing, it should be okay to let the adapters have the request wrapper.

Reply:
@ShriprasadM, we discussed this offline and we're ok with you using the wrapper in analytics adapters given the constraints @hhhjort outlined. As mentioned above, we will need to be on the lookout for attempts to write to the request. I'm going to open up an issue asking for new race tests to help us catch potential writes to the wrapper from within the analytics adapters so it is not just a manual code review effort.